src/runtime_src/core/include/xgq.h

```diff
@@ -215,4 +232,284 @@ struct xrt_cmd_exit_ert {
+
+	/*
+	 * Consumed pointer for both SQ and CQ are here since they don't generate interrupt,
+	 * so no need to be a register.
+	 */
+	uint32_t xh_sq_consumed;
+	uint32_t xh_cq_consumed;
+};
+
+/* Software representation of a single ring buffer. */
+struct xgq_ring {
+	uint32_t xr_slot_num;
+	uint32_t xr_slot_sz;
+	uint32_t xr_produced;
+	uint32_t xr_consumed;
+
+	uint64_t xr_produced_addr;
+	uint64_t xr_consumed_addr;
+	uint64_t xr_slot_addr;
+};
+
+/* Software representation of a single XGQ. */
+struct xgq {
+	bool xq_is_server;
+	uint64_t io_hdl;
+	struct xgq_ring xq_sq ____cacheline_aligned_in_smp;
+	struct xgq_ring xq_cq ____cacheline_aligned_in_smp;
+};
+
+
+/*
+ * XGQ implementation details and helper routines.
+ */
+
+static inline void xgq_copy_to_ring(uint64_t io_hdl, void *buf, uint64_t tgt, size_t len)
+{
+	size_t i;
+	uint32_t *src = (uint32_t *)buf;
+
+	for (i = 0; i < len / 4; i++, tgt += 4)
+		xgq_mem_write32(io_hdl, tgt, src[i]);
+}
+
+static inline void xgq_copy_from_ring(uint64_t io_hdl, void *buf, uint64_t src, size_t len)
+{
+	size_t i;
+	uint32_t *tgt = (uint32_t *)buf;
+
+	for (i = 0; i < len / 4; i++, src += 4)
+		tgt[i] = xgq_mem_read32(io_hdl, src);
+}
+
+static inline void xgq_init_ring(struct xgq_ring *ring, uint64_t produced, uint64_t consumed,
+				 uint64_t slots, uint32_t slot_num, uint32_t slot_size)
+{
+	ring->xr_produced_addr = produced;
+	ring->xr_consumed_addr = consumed;
+	ring->xr_slot_addr = slots;
+	ring->xr_slot_sz = slot_size;
+	ring->xr_slot_num = slot_num;
+	ring->xr_produced = ring->xr_consumed = 0;
+}
+
+static inline bool xgq_ring_full(struct xgq_ring *ring)
+{
+	return (ring->xr_produced - ring->xr_consumed) >= ring->xr_slot_num;
+}
+
+static inline bool xgq_ring_empty(struct xgq_ring *ring)
+{
+	return ring->xr_produced == ring->xr_consumed;
+}
+
+static inline void xgq_ring_read_produced(uint64_t io_hdl, struct xgq_ring *ring)
+{
+	ring->xr_produced = xgq_reg_read32(io_hdl, ring->xr_produced_addr);
+}
+
+static inline void xgq_ring_write_produced(uint64_t io_hdl, struct xgq_ring *ring)
+{
+	xgq_reg_write32(io_hdl, ring->xr_produced_addr, ring->xr_produced);
+}
+
+static inline void xgq_ring_read_consumed(uint64_t io_hdl, struct xgq_ring *ring)
+{
+	ring->xr_consumed = xgq_reg_read32(io_hdl, ring->xr_consumed_addr);
+}
+
+static inline void xgq_ring_write_consumed(uint64_t io_hdl, struct xgq_ring *ring)
+{
+	xgq_reg_write32(io_hdl, ring->xr_consumed_addr, ring->xr_consumed);
+}
+
+static inline uint64_t xgq_ring_slot_ptr(struct xgq_ring *ring, bool produce)
+{
+	uint32_t counter = produce ? ring->xr_produced : ring->xr_consumed;
+
+	return ring->xr_slot_addr + ring->xr_slot_sz * (counter & (ring->xr_slot_num - 1));
+}
+
+static inline bool xgq_can_produce(struct xgq *xgq)
+{
+	struct xgq_ring *ring = xgq->xq_is_server ? &xgq->xq_cq : &xgq->xq_sq;
+
+	if (!xgq_ring_full(ring))
+		return true;
+	xgq_ring_read_consumed(xgq->io_hdl, ring);
+	return xgq_ring_full(ring);
```

Comment:
also need to be return !xgq_ring_full(ring);

Reply:
done